### ISSUE TITLE
Error on accessing prototype's properties

### ISSUE THREAD

**Issue by bgnorlov** [mobxjs/mobx#505]
I am having a very weird error when tried to bind all methods in constructor -"TypeError: Cannot read property 'get' of undefined error inside this.$mobx.values[name].get();. The error only happens when I tried to create second instance of different class
Demo - https://jsbin.com/pevoreyobi/1/edit?html,js,console,output
```
var {observable} = mobx;

class BaseClass {

  @observable requestError = null;

  constructor(){
   var checkObj = this;
    while(checkObj !== Object.prototype){
      Object.getOwnPropertyNames(checkObj).filter(key=>typeof checkObj[key] == 'function' && key!='constructor').forEach(key=>{
        checkObj[key] = checkObj[key].bind(this)
      });
      checkObj = Object.getPrototypeOf(checkObj);
    }
  }
}

class User extends BaseClass {
  @observable select = false;
}

class Board extends BaseClass {
  @observable select = false;
}

var user = new User();
var board = new Board();
```

**Comment by andykog** [comment]
Cleaned up version, representing the bug: https://jsbin.com/wudezul/edit?js,console

**Comment by andykog** [comment]
Cleaned up version, representing the problem: https://jsfiddle.net/Sl1v3r/8f8zmtu5/

@bgnorlov, I think, you shouldn't use `Object.getOwnPropertyNames` for superclass, that you know nothing about. Properties are being made not-enumerable not just for fun. This code will thow as well:
```js
var x = observable([0]);
x.__proto__[0];
```

I wouldn't call this a bug, although fixing this would be nice.

**Comment by andykog** [comment]
Ok, I thought a bit more, and I think it can be called a bug :-)

**Comment by mweststrate** [comment]
I'm very doubtful whether this can be fixed if I understand the problem correctly. The problem here is that the super class uses properties that are not instantiated yet, because super constructores are applied before subclass constructors.

Due to the nature of decorators, this means that the property initializers of the properties are not called yet, although the are already declared on the class. But regardless whether it should throw or not, what you are trying to achieve is imho semantically incorrect.

For example, if you would use extendObservable instead, if would probably not throw, but the BaseClass constructor wouldn't proccess the properties of the subclass in that case (because that constructor is not yet called, so these properties don't exist). Similarly, if you try the fiddle of @andykog code in TypeScript, you will notice that it does not actually throw, (different lifecycle of decorators), but that it also wouldn't work because all subclass values would be undefined.

Note that in any strongly typed language this would actually not be allowed / work correctly; accessing properties of subclasses in the super constructor breaks the contract between super and subclasses (I am not entirely sure whether in java subclass field initializers run before superclass constructors, but at least know that in javascript there are no real field initializers, they are just lifted to assignment statements in the constructor)

I think the only solution to this is to split out this logic in a separate function, and make it the sub-classes responsibility to invoke that at the end of it's constructor function

(on a side note, binding just everything is not a silver bullet, it has it's own hairy downsides; it's expensive and kills the possibility of super calls to name a few)

**Comment by andykog** [comment]
@mweststrate, how about adding checkings in decorator implementation:
```js
get: function() {
  if (this === ObservableObject.prototype) return undefined;
  // ... rest of the getter
}
```
I suppose it can prevent errors from being thrown?

**Comment by mweststrate** [comment]
Not sure, it could help, or add to the confusion. Because this way you
don't get the expected value either, although the behavior could be called
more consistent; always return undefined on not yet initialized properties
instead of throwing. Not sure actually the above check is correct here?
because I think in the example `this` === `instanceof Subclass`?

Op di 23 aug. 2016 om 17:05 schreef Andy Kogut <notifications@github.com>:

**Comment by andykog** [comment]
@mweststrate, oh, you are right

**Comment by mweststrate** [comment]
Hmm we could just xheck if something is there
Op 23 aug. 2016 5:15 p.m. schreef "Andy Kogut" <notifications@github.com>:

**Comment by andykog** [comment]
Yep

**Comment by mweststrate** [comment]
Tried to reproduce, see above commit, but I think I missed the point as these tests succeed

**Comment by mweststrate** [comment]
released candidate `mobx@2.4.3-fix505`, which might fix it, but I didn't get the above JSBin not working (in the babel compilation phase, I have that issue more often?). The possible fix can be found in the above branch

**Comment by andykog** [comment]
You have to uncomment one line to get the error in jsbin
